query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java

```diff
@@ -145,77 +145,79 @@ private Object handleDataWriteCommand(InvocationContext ctx, DataWriteCommand co
       if (command.hasAnyFlag(FlagBitSets.SKIP_INDEXING)) {
          return invokeNext(ctx, command);
       }
-      CacheEntry entry = ctx.lookupEntry(command.getKey());
-      if (ctx.isInTxScope()) {
-         // replay of modifications on remote node by EntryWrappingVisitor
-         if (entry != null && !entry.isChanged() && (entry.getValue() != null || !unreliablePreviousValue(command))) {
-            Map<Object, Object> oldValues = registerOldValues((TxInvocationContext) ctx);
-            oldValues.putIfAbsent(command.getKey(), entry.getValue());
+      return invokeNextThenApply(ctx, command, (rCtx, cmd, rv) -> {
+         if (!cmd.isSuccessful()) {
+            return rv;
          }
+         boolean unreliablePrevious = unreliablePreviousValue(command);
+         if (ctx.isInTxScope()) {
+            Map<Object, Object> oldValues = getOldValuesMap((TxInvocationContext<?>) ctx);
+            registerOldValue(ctx, command.getKey(), unreliablePrevious, oldValues);
+         } else {
+            // TODO: need to reduce the scope of the blocking thread to less if possible later as part of
+            // https://issues.redhat.com/browse/ISPN-11731
+            return asyncValue(blockingManager.runBlocking(() -> indexIfNeeded(rCtx, cmd, unreliablePrevious, cmd.getKey()), cmd)
```

Comment:
since `processChange()` returns a `CompletionStage`, you can change `indexIfNeeded()` to return the `CompletionStage` and void using the `BlockingManager`.

Reply:
makes sense @pruivo , thanks